tls/s2n_cipher_suites.c

```diff
@@ -1168,13 +1216,202 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
+    POSIX_ENSURE_REF(wire);
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(server_index);
+
+    const struct s2n_security_policy *security_policy;
+    POSIX_GUARD(s2n_connection_get_security_policy(conn, &security_policy));
+    const struct s2n_cipher_preferences* cipher_preferences = security_policy->cipher_preferences;
+    POSIX_ENSURE_REF(cipher_preferences);
+
+    bool in_group = false;
+    int negotiated_client_index = count;
+    int negotiated_server_index = -1, negotiated_server_highest_vers_match_index = -1;
+
+    for (uint32_t i = 0; i < cipher_preferences->count; i++) {
+        const struct s2n_cipher_suite* server_cipher = cipher_preferences->suites[i];
+        POSIX_ENSURE_REF(server_cipher);
+
+        /* Check if the cipher suite is an equal-preference grouping delimiter */
+        if (server_cipher == &s2n_equal_preference_group_start) {
+            in_group = true;
+            continue;
+        } 
+        if (server_cipher == &s2n_equal_preference_group_end) {
+            in_group = false;
+            /* Exiting a group and a negotiated cipher has already been found. */
+            if (negotiated_server_index != -1) {
+                *server_index = negotiated_server_index;
+                return S2N_SUCCESS;
+            }
+            continue;
+        }
+
+        /* Cipher suite is NOT a delimiter */
+        int client_index = s2n_wire_ciphers_has_server_cipher_at(server_cipher->iana_value, wire, count, cipher_suite_len);
+
+        /* Client does not support this cipher. Skip. */
+        if (client_index < 0) {
+            continue;
+        }
+
+        /* Found a potential match. Validate that the cipher suite is suitable for this connection. */
+        struct s2n_cipher_suite* match = security_policy->cipher_preferences->suites[i];
+        if (!s2n_cipher_suite_match_is_valid(conn, match)) {
+            continue;
+        }
+
+        /* Don't immediately choose a cipher the connection shouldn't be able to support */
+        if (conn->actual_protocol_version < match->minimum_required_tls_version) {
+            if (negotiated_server_highest_vers_match_index == -1) {
+                negotiated_server_highest_vers_match_index = i;
+            }
+            continue;
+        }
+        if (in_group) {
+            /* Both client and server support a grouped-cipher */
+            if (client_index < negotiated_client_index) {
+                negotiated_client_index = client_index;
+                negotiated_server_index = i;
+            }
+        } else {
+            /* Else, the client and server both support a non-grouped cipher. */
+            negotiated_server_index = i;
+            break;
+        }
+    }
+
+    /* Found a match */
+    if (negotiated_server_index != -1) {
+        *server_index = negotiated_server_index;
+        return S2N_SUCCESS;
+    }
+
+    /* Settle for a cipher with a higher required proto version, if it was set */
```

Comment:
Maybe just explain to me what this case is for? I'm not following.

Reply:
As far as I understand, when we iterate over the offered cipher suites we don't immediately choose a cipher the connection shouldn't be able to support. On the first encounter of a valid cipher that meets the `minimum_required_tls_version`, we store it in a variable and resort to this cipher if another one was not found